packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -959,3 +921,1328 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+// The widget works by inserting its overlay child's render subtree directly
+// under [Overlay]'s render object (_RenderTheatre).
+// https://user-images.githubusercontent.com/31859944/171971838-62ed3975-4b5d-4733-a9c9-f79e263b8fcc.jpg
+//
+// To ensure the overlay child render subtree does not do layout twice, the
+// subtree must only perform layout after both its _RenderTheatre and the
+// [OverlayPortal]'s render object (_RenderLayoutSurrogateProxyBox) have
+// finished layout. This is handled by _RenderDeferredLayoutBox.
+//
+// ** Z-Index of an overlay child
+// [_OverlayLocation] is a (currently private) interface that allows an
+// [OverlayPortal] to insert its overlay child into a specific [Overlay], as
+// well as specifying the paint order between the overlay child and other
+// children of the _RenderTheatre.
+//
+// Since [OverlayPortal] is only allowed to target ancestor [Overlay]s
+// (_RenderTheatre must finish doing layout before _RenderDeferredLayoutBox),
+// the _RenderTheatre should typically be acquired using an [InheritedWidget]
+// (currently, _RenderTheatreMarker) in case the [OverlayPortal] gets
+// reparented.
+
+/// A class to show, hide and bring to top an [OverlayPortal]'s overlay child
+/// in the target [Overlay].
+///
+/// A [OverlayPortalController] can only be given to at most one [OverlayPortal]
+/// at a time. When an [OverlayPortalController] is moved from one
+/// [OverlayPortal] to another, its [isShowing] state does not carry over.
+///
+/// The [show] and [hide] methods typically should not be called when the widget
+/// tree is being rebuilt. However it's allowed to call [show] or [hide] when
+/// the associated [OverlayPortal] widget is yet to be inflated into the widget
+/// tree.
+class OverlayPortalController {
+  /// Creates an [OverlayPortalController].
+  OverlayPortalController({ String? debugLabel }) : _debugLabel = debugLabel;
+
+  _OverlayPortalState? _attachTarget;
+  int? _showTimestamp;
+  final String? _debugLabel;
+
+  static int _wallTime = kIsWeb
+    ? -9007199254740992 // -2^53
+    : -1 << 63;
+
+  // Returns a unique and monotonically increasing timestamp.
+  int _now() {
+    final int now = _wallTime += 1;
+    assert(_showTimestamp == null || _showTimestamp! < now);
+    assert(_attachTarget?._showTimestamp == null || _attachTarget!._showTimestamp! < now);
+    return now;
+  }
+
+  /// Show the overlay child of the [OverlayPortal] this controller is attached
+  /// to, at the top of the target [Overlay].
+  ///
+  /// When there are more than one [OverlayPortal]s that target the same
+  /// [Overlay], the overlay child of the last [OverlayPortal] to have called
+  /// [show] appears at the top level, unobstructed.
+  ///
+  /// If [isShowing] is already true, calling this method brings the overlay
+  /// child it controls to the top.
+  ///
+  /// This method should typically not be called when the widget tree is being
+  /// rebuilt.
+  void show() {
```

Comment:
> you will be able to reuse this controller for multiple overlayportal

This defeats the purpose of having an `OverlayPortalController` (I should add that to the documentation). The timestamp is supposed to be unique (otherwise the group of OverlayPortals would have z-indices that are hard to reason about between each other).

Reply:
Nvm it's already documented: 

/// A [OverlayPortalController] can only be given to at most one [OverlayPortal]
/// at a time. When an [OverlayPortalController] is moved from one
/// [OverlayPortal] to another, its [isShowing] state does not carry over.